File: go/tests/wearable_board_test.go
// wearable_board_test.go
//
// Verifies the CircleAI.Wearable port (wearable_board.go): device add/get +
// vendor-sorted listing, sample recording with unknown-device rejection,
// since-window reads, latest value, and average (NaN when empty).

package circleai_test

import (
	"math"
	"testing"
	"time"

	circleai "github.com/bhengubv/CircleAI/go"
)

func TestWearable_DevicesAndSamples(t *testing.T) {
	b := circleai.NewInMemoryWearableBoard()
	b.Add(circleai.WearableDevice{DeviceId: "d1", Kind: circleai.WearableKindSmartwatch, Vendor: "Zephyr", FirmwareVersion: "1.0", BatteryPct: 80})
	b.Add(circleai.WearableDevice{DeviceId: "d2", Kind: circleai.WearableKindChestStrap, Vendor: "Acme", FirmwareVersion: "2.0", BatteryPct: 50})
	if got, ok := b.GetDevice("d1"); !ok || got.Vendor != "Zephyr" {
		t.Fatalf("get device = %+v ok=%v", got, ok)
	}
	devs := b.Devices()
	if len(devs) != 2 || devs[0].Vendor != "Acme" || devs[1].Vendor != "Zephyr" {
		t.Fatalf("devices sorted by Vendor failed: %+v", devs)
	}

	if err := b.Record(circleai.WearableSample{DeviceId: "ghost", Kind: circleai.WearableTelemetryKindHeartRate, Value: 70, AtUtc: time.Now().UTC()}); err == nil {
		t.Fatalf("recording for unknown device must error")
	}
}

func TestWearable_ReadAndAverage(t *testing.T) {
	b := circleai.NewInMemoryWearableBoard()
	b.Add(circleai.WearableDevice{DeviceId: "d1", Vendor: "Z"})
	base := time.Date(2026, 7, 1, 0, 0, 0, 0, time.UTC)
	must := func(err error) {
		if err != nil {
			t.Fatalf("record: %v", err)
		}
	}
	must(b.Record(circleai.WearableSample{DeviceId: "d1", Kind: circleai.WearableTelemetryKindHeartRate, Value: 60, AtUtc: base}))
	must(b.Record(circleai.WearableSample{DeviceId: "d1", Kind: circleai.WearableTelemetryKindHeartRate, Value: 80, AtUtc: base.Add(time.Hour)}))
	must(b.Record(circleai.WearableSample{DeviceId: "d1", Kind: circleai.WearableTelemetryKindHeartRate, Value: 100, AtUtc: base.Add(2 * time.Hour)}))

	rows := b.ReadSince("d1", circleai.WearableTelemetryKindHeartRate, base.Add(30*time.Minute))
	if len(rows) != 2 || rows[0].Value != 80 || rows[1].Value != 100 {
		t.Fatalf("read-since window failed: %+v", rows)
	}
	if v, ok := b.LatestValue("d1", circleai.WearableTelemetryKindHeartRate); !ok || v != 100 {
		t.Fatalf("latest value = %v ok=%v, want 100", v, ok)
	}
	if _, ok := b.LatestValue("d1", circleai.WearableTelemetryKindSteps); ok {
		t.Fatalf("latest for absent kind must be false")
	}
	if avg := b.AverageValue("d1", circleai.WearableTelemetryKindHeartRate, base); math.Abs(avg-80.0) > 1e-9 {
		t.Fatalf("average = %v, want 80", avg)
	}
	if avg := b.AverageValue("d1", circleai.WearableTelemetryKindSteps, base); !math.IsNaN(avg) {
		t.Fatalf("average with no samples must be NaN, got %v", avg)
	}
}

File: typescript/tests/wearable_board.test.ts
// wearable_board.test.ts
// Verifies the CircleAI.Wearable port: devices sorted by vendor, record guard,
// read-since, latest value, average (NaN when empty), WearableContext record.

import { describe, it } from "node:test";
import assert from "node:assert/strict";
import {
  InMemoryWearableBoard,
  WearableKind,
  WearableTelemetryKind,
  wearableDevice,
  wearableSample,
  wearableContext,
} from "../src/wearable/index";

describe("InMemoryWearableBoard", () => {
  it("lists devices ordered by vendor and enforces enum ordinals", () => {
    const b = new InMemoryWearableBoard();
    b.add(wearableDevice("d1", WearableKind.Smartwatch, "Zephyr", "1.0", 80));
    b.add(wearableDevice("d2", WearableKind.FitnessBand, "Acme", "2.1", 55));
    assert.equal(b.getDevice("d1")?.vendor, "Zephyr");
    assert.deepEqual(
      b.devices.map((d) => d.deviceId),
      ["d2", "d1"], // "Acme" < "Zephyr"
    );
    assert.equal(WearableKind.Smartwatch, 0);
    assert.equal(WearableKind.Headset, 4);
    assert.equal(WearableTelemetryKind.HeartRate, 0);
    assert.equal(WearableTelemetryKind.OxygenPct, 6);
  });

  it("record throws for unknown device", () => {
    const b = new InMemoryWearableBoard();
    assert.throws(
      () => b.record(wearableSample("ghost", WearableTelemetryKind.HeartRate, 70, new Date())),
      /Unknown device ghost/,
    );
  });

  it("reads samples since a cutoff ascending, latest value, and average", () => {
    const b = new InMemoryWearableBoard();
    b.add(wearableDevice("d1", WearableKind.Smartwatch, "Z", "1", 80));
    b.record(wearableSample("d1", WearableTelemetryKind.HeartRate, 60, new Date("2026-01-01T00:00:00Z")));
    b.record(wearableSample("d1", WearableTelemetryKind.HeartRate, 80, new Date("2026-01-03T00:00:00Z")));
    b.record(wearableSample("d1", WearableTelemetryKind.HeartRate, 50, new Date("2025-12-30T00:00:00Z"))); // before cutoff
    const since = new Date("2026-01-01T00:00:00Z");
    assert.deepEqual(
      b.readSince("d1", WearableTelemetryKind.HeartRate, since).map((s) => s.value),
      [60, 80],
    );
    assert.equal(b.latestValue("d1", WearableTelemetryKind.HeartRate), 80);
    assert.equal(b.averageValue("d1", WearableTelemetryKind.HeartRate, since), 70);
  });

  it("latestValue is undefined and averageValue is NaN when there are no samples", () => {
    const b = new InMemoryWearableBoard();
    b.add(wearableDevice("d1", WearableKind.Smartwatch, "Z", "1", 80));
    assert.equal(b.latestValue("d1", WearableTelemetryKind.Steps), undefined);
    assert.ok(Number.isNaN(b.averageValue("d1", WearableTelemetryKind.Steps, new Date("2026-01-01T00:00:00Z"))));
  });

  it("WearableContext carries the biometric snapshot", () => {
    const ctx = wearableContext(72, 4200, 98, 36.5, true, new Date("2026-01-01T00:00:00Z"));
    assert.equal(ctx.heartRateBpm, 72);
    assert.equal(ctx.stepCountToday, 4200);
    assert.equal(ctx.isWorkoutActive, true);
    const empty = wearableContext(null, null, null, null, false, new Date("2026-01-01T00:00:00Z"));
    assert.equal(empty.spO2Percent, null);
  });
});
